lib/matplotlib/image.py

```diff
@@ -162,203 +210,158 @@ def changed(self):
+        magnified by the magnification factor.
+
+        If `unsampled` is True, the image will not be scaled, but an
+        appropriate affine transformation will be returned instead.
+
+        If `round_to_pixel_border` is True, the output image size will
+        be rounded to the nearest pixel boundary.  This makes the
+        images align correctly with the axes.  It should not be use in
+        cases where you want exact scaling, however, such as
+        FigureImage.
+
+        Returns the resulting (image, x, y, trans), where (x, y) is
+        the upper left corner of the result in pixel space, and
+        `trans` is the affine transformation from the image to pixel
+        space if the image data itself was not resampled.
         """
-        xmin, xmax, ymin, ymax = image_extents
-        dxintv = xmax-xmin
-        dyintv = ymax-ymin
+        if A is None:
+            raise RuntimeError('You must first set the image'
+                               ' array or the image attribute')
 
-        # the viewport scale factor
-        if viewlim.width == 0.0 and dxintv == 0.0:
-            sx = 1.0
-        else:
-            sx = dxintv/viewlim.width
-        if viewlim.height == 0.0 and dyintv == 0.0:
-            sy = 1.0
-        else:
-            sy = dyintv/viewlim.height
-        numrows, numcols = A.shape[:2]
-        if sx > 2:
-            x0 = (viewlim.x0-xmin)/dxintv * numcols
-            ix0 = max(0, int(x0 - self._filterrad))
-            x1 = (viewlim.x1-xmin)/dxintv * numcols
-            ix1 = min(numcols, int(x1 + self._filterrad))
-            xslice = slice(ix0, ix1)
-            xmin_old = xmin
-            xmin = xmin_old + ix0*dxintv/numcols
-            xmax = xmin_old + ix1*dxintv/numcols
-            dxintv = xmax - xmin
-            sx = dxintv/viewlim.width
-        else:
-            xslice = slice(0, numcols)
-
-        if sy > 2:
-            y0 = (viewlim.y0-ymin)/dyintv * numrows
-            iy0 = max(0, int(y0 - self._filterrad))
-            y1 = (viewlim.y1-ymin)/dyintv * numrows
-            iy1 = min(numrows, int(y1 + self._filterrad))
-            yslice = slice(iy0, iy1)
-            ymin_old = ymin
-            ymin = ymin_old + iy0*dyintv/numrows
-            ymax = ymin_old + iy1*dyintv/numrows
-            dyintv = ymax - ymin
-            sy = dyintv/viewlim.height
-        else:
-            yslice = slice(0, numrows)
+        clipped_bbox = Bbox.intersection(out_bbox, clip_bbox)
 
-        if xslice != self._oldxslice or yslice != self._oldyslice:
-            self._imcache = None
-            self._oldxslice = xslice
-            self._oldyslice = yslice
+        if clipped_bbox is None:
+            return None, 0, 0, None
 
-        if self._imcache is None:
-            A = self._A
-            if self.origin == 'upper':
-                A = A[::-1]
+        out_width_base = clipped_bbox.width * magnification
+        out_height_base = clipped_bbox.height * magnification
```

Comment:
It shouldn't ever be negative, but it comes (ultimately) from the backend renderer's `get_image_magnification` method (which depends on `dpi` which is checked to be positive).  So I don't think a check is necessarily in order, given that this is a private method.